ibm/resource_ibm_ipsec_vpn.go

```diff
@@ -0,0 +1,466 @@
+	log.Printf("[INFO] IPSec VPN ID: %s", d.Id())
+	return resourceIBMIPSecVPNUpdate(d, meta)
+}
+
+func findIPSecVpnByOrderID(sess *session.Session, orderID int, d *schema.ResourceData) (datatypes.Network_Tunnel_Module_Context, error) {
+	filterPath := "networkTunnelContexts.billingItem.orderItem.order.id"
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"pending"},
+		Target:  []string{"complete"},
+		Refresh: func() (interface{}, string, error) {
+			vpn, err := services.GetAccountService(sess).
+				Filter(filter.Build(
+					filter.Path(filterPath).
+						Eq(strconv.Itoa(orderID)))).
+				Mask(ipsecMask).
+				GetNetworkTunnelContexts()
+			if err != nil {
+				return datatypes.Network_Tunnel_Module_Context{}, "", err
+			}
+
+			if len(vpn) == 1 {
+				return vpn[0], "complete", nil
+			} else if len(vpn) == 0 {
+				return nil, "pending", nil
+			}
+			return nil, "", fmt.Errorf("Expected one IPSec VPN: %s", err)
+		},
+		Timeout:        2 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     10 * time.Second,
+		NotFoundChecks: 24 * 60,
+	}
+
+	pendingResult, err := stateConf.WaitForState()
+
+	if err != nil {
+		return datatypes.Network_Tunnel_Module_Context{}, err
+	}
+	var result, ok = pendingResult.(datatypes.Network_Tunnel_Module_Context)
+	if ok {
+		return result, nil
+	}
+
+	return datatypes.Network_Tunnel_Module_Context{},
+		fmt.Errorf("Cannot find IPSec Vpn with order id '%d'", orderID)
+}
+
+func resourceIBMIPSecVPNRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	vpn, err := services.GetNetworkTunnelModuleContextService(sess).
+		Id(vpnID).Mask(ipsecMask).
+		GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving firewall information: %s", err)
+	}
+	d.Set("name", *vpn.Name)
+	d.Set("internal_peer_ip_address", *vpn.InternalPeerIpAddress)
+	return nil
+}
+
+func resourceIBMIPSecVPNExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	sess := meta.(ClientSession).SoftLayerSession()
+
+	fwID, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return false, fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	_, err = services.GetNetworkTunnelModuleContextService(sess).
+		Id(fwID).
+		GetObject()
+
+	if err != nil {
+		if apiErr, ok := err.(sl.Error); ok && apiErr.StatusCode == 404 {
+			return false, nil
+		}
+		return false, fmt.Errorf("Error retrieving vpn information: %s", err)
+	}
+
+	return true, nil
+}
+
+func resourceIBMIPSecVPNDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnService := services.GetNetworkTunnelModuleContextService(sess)
+
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	// Get billing item associated with the firewall
+	billingItem, err := vpnService.Id(vpnID).GetBillingItem()
+
+	if err != nil {
+		return fmt.Errorf("Error while looking up billing item associated with the firewall: %s", err)
```

Comment:
Log stmt needs to be corrected from firewall to ipsecvpn

Reply:
Done. Please review